File: esp32/sonos-controller-client.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define ROWS 5
#define COLS 24
#define INITVOL 50
#define SERVER_ADDRESS "192.168.0.223"
#define SERVER_PORT "5001"

class CommDriver {
  // This will talk to the sonos server hosted elsewhere
  private:
    int rows=ROWS;
  public:
    CommDriver() {
      // Use 'this->' to make the difference between the
      // 'pin' attribute of the class and the 
      // local variable 'pin' created from the parameter.
      init();
    }

    void init() {
        return;
    }
    void start_playlist(){
      printf("start playlist");
    }
    void playpause(){
      printf("play pause");
    }
    void next(){
      printf("next");
    }
    void prev(){
      printf("prev");
    }
}; // don't forget the semicolon at the end of the class


class ScreenDriver {
  // This will manage the tft with TTGO specific call
  private:
    int rows=ROWS;
    int cols=COLS;
    char prefixes[5][5]={ "","","","",""};
    char values[5][256]={ "","","","",""};
  public:
    ScreenDriver() {
      // Use 'this->' to make the difference between the
      // 'pin' attribute of the class and the 
      // local variable 'pin' created from the parameter.
      init();
    }

    void init() {
        return;
    }
    void setprefix(int idx, char *val){
      strcpy(prefixes[idx],val);      
    }
    void setvalue(int idx, char *val){
      strcpy(values[idx],val);      
    }

    void render(){
      for (int n=0;n<5;n++){
        printf("%s%s\n",prefixes[n],values[n]);        
      }      
    }
}; // don't forget the semicolon at the end of the class



class ScreenChooser {
  private:
    int pos=0;
    char prefixes[ROWS][5]={ "---:","---:",">>>:","---:","---:"};
    char values[ROWS][256]={ "","","","",""};
    char all_playlists[10][256] = {"","","queen","ac/dc","bowie","robbie gill","turkuaz","the daily","",""};
    ScreenDriver *sd;
    CommDriver *cd;
  public:
    long int id=0;
    ScreenChooser(CommDriver* cd1, ScreenDriver* sd1) {
      // Use 'this->' to make the difference between the
      // 'pin' attribute of the class and the 
      // local variable 'pin' created from the parameter.
      sd=sd1;
      cd=cd1;
      init();
    }
    void init() {
        id=random();
        return;
    }
    void scroll(int increment){
      int newpos=pos+increment;
      if(  newpos >= 0 && newpos <= sizeof(all_playlists)){
        pos=newpos;
      }
    }
    int scroll_select(){
      return pos;
    }    
    void render(){
      int i;
      int plidx;
      for (i=0;i<ROWS;i++){
        sd->setprefix(i,prefixes[i]);
      }
      for (i=0;i<ROWS;i++){
        plidx=pos+i;
        sd->setvalue(i,all_playlists[plidx]);
      }
      sd->render();
    }

}; // don't forget the semicolon at the end of the class


class ScreenDefault {
  private:
    int vol=INITVOL;
    char prefixes[ROWS][5]={ "    ","Vol:","Now:","PL :","   "};
    char values[ROWS][256]={ "","","","",""};
    ScreenDriver *sd;
    CommDriver *cd;
  public:
    long int id=0;
    ScreenDefault(CommDriver* cd1, ScreenDriver* sd1) {
      // Use 'this->' to make the difference between the
      // 'pin' attribute of the class and the 
      // local variable 'pin' created from the parameter.
      sd=sd1;
      cd=cd1;
      init();
    }
    void init() {
        id=random();
        return;
    }
    void scroll(int increment){
      int newvol=vol+increment;
      if(  newvol >= 0 && newvol  <=100){
        vol=newvol;
      }
    }
    void scroll_select(){
      return;
    }    
    void playpause(){
      cd->playpause();
      return;
    }    
    void next(){
      cd->next();
      return;
    }    
    void prev(){
      cd->prev();
      return;
    }    
    void set_volstring(char *c){
      strcpy(c,"[ volstring ]");
    }    
    void set_playingstring(char *c){
      strcpy(c,"playingstring");
    }    
    void render(){
      int i;
      set_volstring(values[1]);
      set_playingstring(values[2]);
      for (i=0;i<ROWS;i++){
        sd->setprefix(i,prefixes[i]);
      }
      for (i=0;i<ROWS;i++){
        sd->setvalue(i,values[i]);
      }
      sd->render();
    }

}; // don't forget the semicolon at the end of the class

class SonosApp {
  private:
    int rows=ROWS;
    //0=default, 1=chooser
    int current=0;
    ScreenChooser* sc;
    ScreenDefault* sd;
  public:
    SonosApp(ScreenDefault* sd1, ScreenChooser* sc1) {
      // Use 'this->' to make the difference between the
      // 'pin' attribute of the class and the 
      // local variable 'pin' created from the parameter.
      sd=sd1;
      sc=sc1;
      init();
    }

    void init() {
        return;
    }
    void scroll_select(){
      if (current == 0 ){
        current=1;
        sc->render();        
      }
      else if (current==1){
        current=0;
        sd->render();
      }
    }
    void scroll(int n){
      if (current ==0 ){
        //
      }
      else if (current == 1){
        sc->scroll(n);
      }
    }

    void render(){
      if (current ==0 ){
        sd->render();
      }
      else if (current == 1){
        sc->render();
      }
    }
}; // don't forget the semicolon at the end of the class


CommDriver cd2;
int main(void){
    CommDriver cd;
    ScreenDriver sd;
    ScreenDefault sdd(&cd,&sd);
    ScreenChooser sc(&cd,&sd);
    SonosApp sa(&sdd,&sc);
    sa.render();
    printf("\n");
    sa.scroll_select();
    sa.render();
    printf("\n");
    sa.scroll(1);
    sa.render();
    printf("\n");
    
}

File: esp32/sonos-controller-client.py
#!/usr/bin/python3
import os
import time
import sys
PLAYLISTS=["queen","ac/dc","bowie","robbie gill","turkuaz","the daily"]
ROWS=5
WIDTH=24
INITIAL_VOLUME=50
DEFAULT_TARGET="192.168.0.223"

class CommDriver:
    """Talk to the sonos controller service"""
    def __init__(self,target=DEFAULT_TARGET):
        pass
    def start_playlist(self,pl):
        print(f"start playlist {pl}")

    def play_pause(self):
        print(f"play/pause")
    def next(self):
        print(f"next")
    def prev(self):
        print(f"prev")

class ScreenDriver:
    """Consistently render the screen"""
    def __init__(self, rows=ROWS, cols=WIDTH):
        self.rmax=rows
        self.wmax=cols
        self.prefixes=["","","","",""]
        self.values=["","","","",""]

    def set_prefix(self,n,val):
        self.prefixes[n]=val

    def set_value(self,n,val):
        self.values[n]=val

    def render(self):
        for n in range(0,self.rmax):
            print(f"{self.prefixes[n]}{self.values[n]}")


class ScreenDefault:
    """The home screen for the app"""
    def __init__(self, sd:ScreenDriver ,cd=CommDriver, initial_volume=INITIAL_VOLUME,playing="Evening dinner",width=24):
        self.sd=sd
        self.cd=cd
        self.width=width
        self.max_vol=100
        self.vol=initial_volume
        self.playing=playing
        self.prefixes=["","Vol: ","Now: ","PL : ",""]
        self.values=["","","","",""]

    @property
    def vol_string(self):
        max_size = self.width - len(self.prefixes[0]) - 2
        chars = int((self.vol / 100) * max_size)
        n=0
        str=""
        while(n<max_size):
            if n<chars:
                str=str+"="
            else:
                str=str+" "
            n=n+1
        return f"[{str}]"

    @property
    def playing_string(self):
        return self.playing

    def scroll(self,inc):
        if (self.vol + inc) >= 0 and (self.vol + inc) <= self.max_vol:
            self.vol = self.vol + inc

    def scroll_select(self):
        pass

    def set_playing(self,new_playing):
        self.playing=new_playing

    def play_pause(self):
        self.cd.play_pause()
    def next(self):
        self.cd.next()
    def prev(self):
        self.cd.prev()

    def render(self):
        os.system('clear')
        self.values[1]=self.vol_string
        self.values[2]=self.playing_string
        for n,v in enumerate(self.prefixes):
            self.sd.set_prefix(n,v)
        for n,v in enumerate(self.values):
            self.sd.set_value(n,v)
        self.sd.render()

def test_default():
    sd=ScreenDriver()
    x=ScreenDefault(sd)
    x.render()

    for n in range(0,11):
        x.scroll(10)
        time.sleep(0.2)
        x.render()

    for s in ["ac/dc","bowie","turkuaz"]:
        x.set_playing(s)
        x.render()
        time.sleep(0.2)

    for n in range(0,11):
        x.scroll(-10)
        time.sleep(0.2)
        x.render()


class ScreenChooser:
    """Select from playlists"""
    def __init__(self, sd:ScreenDriver, cd:CommDriver, rmax=ROWS,all_playlists=PLAYLISTS):
        self.sd=sd
        self.cd=cd
        self.all_playlists=all_playlists
        self.rows=[]
        self.rmax=rmax
        self.pos=0
        self.prefixes=["---:","---:",">>>:","---:","---:"]

    
    def set_all_playlists(self,ap):
        self.all_playlists=ap

    def scroll(self,inc):
        if (self.pos + inc) >= 0 and (self.pos + inc) < len(self.all_playlists):
            self.pos = self.pos + inc

    def scroll_select(self):
        self.cd.start_playlist(self.selected)
        return self.selected

    @property
    def rowlist(self):
        rows=[]
        rows.extend(["",""])
        rows.extend(self.all_playlists)
        rows.extend(["",""])
        out= rows[self.pos:self.pos+self.rmax]
        self.selected=out[3]
        return out

    def render(self):
        os.system('clear')
        for n,v in enumerate(self.prefixes):
            self.sd.set_prefix(n,v)
        for n,v in enumerate(self.rowlist):
            self.sd.set_value(n,v)
        self.sd.render()

def test_screen_chooser():
    sd=ScreenDriver()
    cd=CommDriver()
    x=ScreenChooser(sd,cd)
    x.set_all_playlists(sorted(["queen","ac/dc","bowie","robbie gill","turkuaz","the daily"]))
    x.render()

    for n in range(0,10):
        time.sleep(0.3)
        x.scroll(1)
        x.render()

    for n in range(0,10):
        time.sleep(0.3)
        x.scroll(-1)
        x.render()

class App:
    """Control the app"""
    def __init__(self, screen_default:ScreenDefault, screen_chooser:ScreenChooser):
        self.sd=screen_default
        self.sc=screen_chooser
        self.current="default"
        self.ss = screen_default

    def scroll_action(self,inc):
        self.ss.scroll(inc)
        self.ss.render()

    def scroll_select(self):
        val=self.ss.scroll_select()
        if self.current=="default":
            self.current="chooser"
            self.ss=self.sc
        elif self.current=="chooser":
            self.current="default"
            self.ss=self.sd
            self.ss.set_playing(val)
        self.ss.render()

    def pp(self):
        self.sd.play_pause()
    def prev(self):
        self.sd.prev()
    def next(self):
        self.sd.next()
    def render(self):
        self.ss.render()

def test_app():
    sd=ScreenDriver()
    comm=CommDriver()
    x=App(ScreenDefault(sd,comm),ScreenChooser(sd,comm))
    x.render()
    while(1):
        action=input(":").lower()
        if action.startswith("l"):
            x.scroll_action(-1)
        if action.startswith("p"):
            x.scroll_action(1)
        if action.startswith("o"):
            x.scroll_select()
        if action.startswith("z"):
            x.pp()
        if action.startswith("x"):
            x.prev()
        if action.startswith("c"):
            x.next()
        if action.startswith("q"):
            sys.exit()

test_default()
time.sleep(1)
test_screen_chooser()
test_app()
sys.exit()

